docs/investigation/configuration/2022-09-07-persistent-cfg-from-commands.md

```diff
@@ -0,0 +1,418 @@
+
+### 2. Restart app every time and post an updated message 
+
+Every command, even `notifier start/stop` restarts the app. But we don't post "config change detected", "goodbye" and "hello" messages.
+
+Instead, we post a notification when BotKube is online: `BotKube configuration for cluster "dev" has been reloaded 👍`.
+
+How to implement it? There would be another "state" BotKube ConfigMap, which is not monitored by Config Watcher, but still loaded during BotKube startup. Let's call it "startup-state" for now.
+
+When detecting new configuration, Config Watcher writes to the "startup-state":
+
+```yaml
+lastExitReason: "ConfigReloaded"
+```
+
+When BotKube launches again, it reads the config and posts `BotKube configuration for cluster "dev" has been reloaded 👍`. Then, it modifies the "startup-state" ConfigMap and removes the `lastExitReason` field, to not post the same message next time.
+
+Initially, it is posted for all channels as it is right now. Later, we can include the channel name in the state and post it only to the channel where the command was executed.
+
+Pros:
+- Quite easy to implement and maintain
+- BotKube reload is "hidden"
+
+Cons:
+- Current `notifier start/stop` commands will also reload BotKube. That shouldn't make big difference though, as it would be "hidden".
+
+### 3. Distinguish two types of commands and restart if necessary
+
+This is a variation of the previous approach. We could distinguish two types of commands:
+- commands that don't restart BotKube, e.g. `notifier start/stop`
+  - they save the state in runtime config and another ConfigMap, which is not monitored by Config Watcher, but still loaded during BotKube startup.
+- commands that restart BotKube instantly, such as commands for notification presets
+  - they post "Reloading configuration..." message and restart the app.
+  
+Initially, the "goodbye" and "welcome" messages could be kept as they are. Later we can combine it with option 2.
+
+Pros:
+- Quite easy to implement and maintain
+- Current `notifier start/stop` still work instantly
+
+Cons:
+- If we want to combine it with option 2 approach to hide BotKube restart, it would be a bit more time-consuming to implement.
+
+### Rejected: Reload config dynamically
+
+This doesn't change much as basically it would be restarting all components without the app itself; there would be still the same issues as with current approach. Kubernetes spawns a new Pod instantly, so there shouldn't be much time difference.
+It would complicate the code though (as we would need to have another level of abstraction to watch over full app and ensure it's not finished, but restarted instead).
+
+Doing a diff and restarting just some updated components (e.g. just Slack Bot or whole source router) doesn't sound like something we'd like to implement and maintain, as it'll bring too much complexity into our code, and, in a result, unpredictability in the behavior.
+
+## Consequences
+
+TBD - once we decide on the approach.
```

Comment:
This is what I wanted to write down if everyone agrees:

```suggestion
## Decision

After team discussion (@ezodude, @huseyinbabal, @mszostok) we agree as follows:
- We'll choose the option no 3. In the first implementation every BotKube restart related to configuration change will post "goodbye" and "hello" messages. This will be implemented as a part of [#704](https://github.com/kubeshop/botkube/issues/704).
- We may still implement the `@BotKube reload` command from the [1. Dedicated command for manual restart](#1-dedicated-command-for-manual-restart) section later, to support operation use cases (e.g. update configuration during maintenance window). This will be defined as a follow-up task to see how big the community demand is.
```

Reply:
Modified the suggestion after discussion with @mszostok - added scope of 2 as a stretch 👍

Reply:
That's totally fine by me, thanks for the effort and collaboration 🎖️